x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -101,9 +104,9 @@ protected List<Batch<PhysicalPlan>> rules(boolean optimizeForEsSource) {
         esSourceRules.add(new ReplaceAttributeSourceWithDocId());
 
         if (optimizeForEsSource) {
-            esSourceRules.add(new PushTopNToSource());
+            esSourceRules.add(new PushTopNToSource(context()));
```

Comment:
Since these two rules have similar logic it's worth making the rules non-static, move hasIdenticalDelegate as a private method to the outer class and have the classes refer to that.
canPushDown will still have to use the static signature however it should be less bureaucratic to access the `SearchStats`.

Reply:
Scratch the above - use the same pattern as `ReplaceMissingFieldWithNull` in `LocalLogicalPlanOptimizier` that is change the rule to be a `ParameterizedRule` so the context will be passed in per method instead inside the constructor.